R/qenv-extract.R

```diff
@@ -0,0 +1,48 @@
+#'
+#' @section Subsetting:
+#' `x[names]` subsets objects in `qenv` environment and limit the code to the necessary needed to build limited objects.
+#'
+#' @param names (`character`) names of objects included in `qenv` to subset
+#' @param x (`qenv`)
+#'
+#' @examples
+#'
```

Comment:
Would it make sense to expand the subset to include all objects that are related (by code dependency)

See pseudo-test below.

This creates an inconsistency where the code and environment doesn't match. ([issue in teal](https://github.com/insightsengineering/teal/issues/1390))

``` r
pkgload::load_all("/teal.code")
#> ℹ Loading teal.code
q <- qenv() |> within({
  a <- 2
  b <- a + 1
})

q2 <- q["b"]

get_code(q2) |> paste(collapse = "\n") |> cat()
#> a <- 2
#> b <- a + 1
expect_identical(get_code(q2), c("a <- 2", "b <- a + 1"))

ls(get_env(q2))
#> [1] "b"
expect_setequal(ls(get_env(q2)), c("a", "b"))
#> Error: ls(get_env(q2)) (`actual`) and c("a", "b") (`expected`) don't have the same values.
#> * Only in `actual`: "a"
```

<sup>Created on 2024-10-30 with [reprex v2.1.1](https://reprex.tidyverse.org)</sup>

Reply:
Oh yeah, that's very very needed

Reply:
@averissimo you would probably not like to extract `a` in case `a` is modified after `b`

```r
q <- qenv() |> within({
  a <- 2
  b <- a + 1
  a <- a + 2
})
```

What is the point of having `a` extracted in `@env` for `q["b"]` in this case? 
We just need `b` and the code needed to reproduce `b`.

Reply:
That's a really good point and my only thought is to simplify, instead of adding complexity.

We could "cheat" by removing non-subsetted objects, let's say in pseudo-code

``` r
q <- qenv() |> within({
  a <- 2
  b <- a + 1
  a <- a + 2
})

q2 <- q["b"]

q2 |> get_code() |> cat()
#> a <- 2
#> b <- a + 1
#> a <- a + 2
#> rm(a)
```


Subsetting the objects to get all related dependencies sounds good on paper in general, and it makes sense when considering `qenv` in isolation.

Upstream on teal, the modules must be specific and be included on the `datanames` for it to be used on the modules. So that issue may be closed.